### ISSUE TITLE
Fix Windows build

### ISSUE THREAD

**Issue by XVilka** [fxfactorial/ocaml-linenoise#15]
Currently it's marked as incompatible by @fdopen: https://github.com/fdopen/opam-repository-mingw/commit/23748b9f38c55dd679987911fe21e2b6f4537abf

It's required for enabling Windows builds for BAP: https://github.com/BinaryAnalysisPlatform/bap/pull/1254

**Comment by c-cube** [comment]
What breaks on windows? I don't have a windows machine on hand, so help would be appreciated.

**Comment by ivg** [comment]
Here are the details, https://github.com/antirez/linenoise/issues/113 it looks like that there are workarounds, i.e., versions of linenoise that we can compiler under windows (referenced in the issue). See also https://github.com/ocaml-community/ocaml-linenoise/blob/master/src/linenoise_src.c#L49

Also, I think that it would be OK, at least for starters, to just fallback dump input on windows.

**Comment by c-cube** [comment]
> Also, I think that it would be OK, at least for starters, to just fallback dump input on windows.

I'm swamped right now (new baby!!) but if you're able to contribute that it'd be awesome.

**Comment by c-cube** [comment]
I think using [redox liner](https://gitlab.redox-os.org/redox-os/liner) for example would be better on the long term. It's more manageable than a C codebase, especially wrt unicode.

**Comment by ivg** [comment]
But this will require the presence of a rust compiler/package manager and will make packing of linenoise a bit convoluted.

**Comment by c-cube** [comment]
It would require cargo, yes. The rest of the packaging wouldn't change much (you can vendor the entire tree of the project).

**Comment by ivg** [comment]
Yep, and requiring cargo is a much bigger hassle than requiring a C compiler (which is already required by opam itself). I don't really think that unfolding a new full toolchain is justified for linenoise which is considered as a lightweight drop in tool. Also, the name itself will no longer be correct :) I mean having a different readline-like utility with the redox backing is probably a good idea. But switching from linenoise to redox as the main backend for ocaml-linenoise doesn't sound that much good and will prevent at least myself from using it. 

Besides, I am still planning to resolve this issue, just finishing stuff from the last year. My main plan is to adopt one of the portability clones of linenoise referenced in antirez/linenoise#113

It doesn't look like a lot of work, except that it was more than 20 years ago when I built something on Windows)

**Comment by c-cube** [comment]
If you think you can fix the C, then power to you!! :grin:

**Comment by pmetzger** [comment]
@ivg Do you want to take a crack at it?

**Comment by StDymphna** [comment]
It's a bit of a mess since it looks like they're all based on different versions of the upstream linenoise, but there is a working fork of linenoise with win32 support (https://github.com/msteveb/linenoise), with some missing/fixed features in a yet to be added pull request here: https://github.com/msteveb/linenoise/pull/27 . It was possible for me to get ocaml-linenoise to easily compile using this fork.

**Comment by c-cube** [comment]
@EmmaJaneBonestell just to make it clear, does this fork also handle the features we added, like ctrl-r? At a glance it seems like it but I'm not sure.

**Comment by StDymphna** [comment]
@c-cube With the pull request, I believe so, yes. I haven't done any extensive testing, but it seems to work all right.  It seems like it would be easier to implement/lift needed Windows code from this.  The fork added lots of fixes/cleanups like 11 and 9 years ago, and I don't know if they are still relevant, should be kept, etc. (e.g. escaping in linenoiseHistoryLoad.) freehistory is named linenoiseHistoryFree and has two more lines of code for some supposed bug fixes, and so on.

For anyone who comes across this and needs a working repo (not guaranteed to match all features/be bug free):
https://github.com/EmmaJaneBonestell/ocaml-linenoise

**Comment by ivg** [comment]
@EmmaJaneBonestell, would you mind making a pull request to this repository?
